File: rxgooglemap/src/main/java/com/sdoward/rxgooglemap/CircleClickFunc.java
package com.sdoward.rxgooglemap;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.Circle;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.functions.Function;

class CircleClickFunc implements Function<GoogleMap, Observable<Circle>> {

    @Override
    public Observable<Circle> apply(final GoogleMap googleMap) {
        return Observable.create(new ObservableOnSubscribe<Circle>() {
            @Override
            public void subscribe(final ObservableEmitter<Circle> emitter) throws Exception {
                googleMap.setOnCircleClickListener(new GoogleMap.OnCircleClickListener() {
                    @Override
                    public void onCircleClick(Circle circle) {
                        emitter.onNext(circle);
                    }
                });
            }
        });
    }
}

File: rxgooglemap/src/main/java/com/sdoward/rxgooglemap/CircleClickFunc.kt
package com.sdoward.rxgooglemap

import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.Circle
import io.reactivex.Observable
import io.reactivex.functions.Function

class CircleClickFunc : Function<GoogleMap, Observable<Circle>> {

    override fun apply(t: GoogleMap): Observable<Circle> {
        return Observable.create { subscriber ->
            t.setOnCircleClickListener {
                subscriber.onNext(it)
            }
        }
    }

}